cli/parse.go

```diff
@@ -341,29 +265,142 @@ func parseArgs(inputs []string, stdin *os.File, argDefs []cmdkit.Argument, recur
 				argDef.Required && !fillingVariadic {
 				r, err := maybeWrapStdin(stdin, msgStdinInfo)
 				if err != nil {
-					return nil, nil, err
+					return err
 				}
 
 				fpath := stdin.Name()
 				fileArgs[fpath] = files.NewReaderFile("", fpath, r, nil)
 			}
 		}
 
-		argDefIndex++
+		iArgDef++
+	}
+
+	if iArgDef == len(argDefs)-1 && stdin != nil &&
+		req.Command.Arguments[iArgDef].SupportsStdin {
+		// handle this one at runtime, pretend it's there
+		iArgDef++
 	}
 
 	// check to make sure we didn't miss any required arguments
-	if len(argDefs) > argDefIndex {
-		for _, argDef := range argDefs[argDefIndex:] {
+	if len(argDefs) > iArgDef {
+		for _, argDef := range argDefs[iArgDef:] {
 			if argDef.Required {
-				return nil, nil, fmt.Errorf("Argument '%s' is required", argDef.Name)
+				return fmt.Errorf("argument %q is required", argDef.Name)
+			}
+		}
+	}
+
+	req.Arguments = stringArgs
+	if len(fileArgs) > 0 {
+		req.Files = files.NewSliceFile("", "", filesMapToSortedArr(fileArgs))
+	}
+
+	return nil
+}
+
+func splitkv(opt string) (k, v string, ok bool) {
+	split := strings.SplitN(opt, "=", 2)
+	if len(split) == 2 {
+		return split[0], split[1], true
+	} else {
+		return opt, "", false
+	}
+}
+
+func parseOpt(opt, value string, opts map[string]cmdkit.Option) interface{} {
+	optDef, ok := opts[opt]
+	if !ok {
+		panic(fmt.Errorf("unknown option %q", opt))
+	}
+
+	v, err := optDef.Parse(value)
+	if err != nil {
+		panic(err)
+	}
+	return v
+}
+
+type kv struct {
+	Key   string
+	Value interface{}
+}
+
+func parseShortOpts(cmdline []string, i int, optDefs map[string]cmdkit.Option) ([]kv, int) {
+	k, vStr, ok := splitkv(cmdline[i][1:])
+	var (
+		j   int
```

Comment:
why declare j here? could just do it down in the for loop, no?

Reply:
no good reason. I put it into the for header now.